src/query_param.c

```diff
@@ -38,27 +38,21 @@ QueryParam *NewNumericFilterQueryParam_WithParams(struct QueryParseCtx *q, Query
   return ret;
 }
 
-QueryParam *NewVectorFilterQueryParam_WithParams(struct QueryParseCtx *q, QueryToken *vec, QueryToken *type, QueryToken *value) {
+// TODO: to be more generic, consider using variadic function, or use different functions for each command
+QueryParam *NewVectorFilterQueryParam_WithParams(struct QueryParseCtx *q, VectorQueryType type, QueryToken *value, QueryToken *vec) {
   QueryParam *ret = NewQueryParam(QP_VEC_FILTER);
   VectorFilter *vf = rm_calloc(1, sizeof(*vf));
-  VectorFilter_InitValues(vf);
   ret->vf = vf;
-  int prevNumParams = q->numParams;
-  QueryParam_InitParams(ret, 3);
-  QueryParam_SetParam(q, &ret->params[0], &vf->vector, &vf->vecLen, vec);
-  assert (type->type != QT_TERM_CASE);
-  if (type->type == QT_TERM) {
-    vf->type = VectorFilter_ParseType(type->s, type->len);
-  } else {
-    QueryParam_SetParam(q, &ret->params[1], &vf->type, NULL, type);
-  }
-  QueryParam_SetParam(q, &ret->params[2], &vf->value, NULL, value);
-  if (prevNumParams == q->numParams) {
-    // No parameters were used - need to validate now
-    if (!VectorFilter_Validate(vf, q->status)) {
+  vf->type = type;
+  switch (type) {
+    case VECSIM_QT_TOPK:
+      QueryParam_InitParams(ret, 2);
+      QueryParam_SetParam(q, &ret->params[0], &vf->topk.value, NULL, value);
+      QueryParam_SetParam(q, &ret->params[1], &vf->topk.vector, &vf->topk.vecLen, vec);
+      break;
+    default:
```

Comment:
you need to free `vf` also on error.

Reply:
we set `vf` to be `ret`'s `vf`, and we free it inside `QueryParam_Free(ret);` on error